File: lab3/src/main/scala/ru/ifmo/ctddev/komarov/net/lab3/state/BroadcastMessage.scala
package ru.ifmo.ctddev.komarov.net.lab3.state

import ru.ifmo.ctddev.komarov.net.lab3.crypto.elgamal.PublicKey
import ru.ifmo.ctddev.komarov.net.lab3.crypto.SHA256Hash

case class BroadcastMessage(x: PublicKey, revList: SHA256Hash) {
  def getBytes: Array[Byte] = x.getBytes ++ revList.bytes
}

object BroadcastMessage {
  private val revListStart = PublicKey.BYTE_LENGTH
  private val end = revListStart + SHA256Hash.BYTE_LENGTH

  def fromBytes(arr: Array[Byte]): Option[BroadcastMessage] = {
    if (arr.length != PublicKey.BYTE_LENGTH) {
      None
    } else {
      for (
        k <- PublicKey(arr.slice(0, revListStart));
        h <- Some(SHA256Hash(arr.slice(revListStart, end)))
      ) yield BroadcastMessage(k, h)
    }
  }
}

File: lab3-java/src/lab3/structs/BroadcastMessage.java
package lab3.structs;

import lab3.bytes.Storable;
import lab3.crypto.SHA256Hash;
import lab3.crypto.elgamal.PublicKey;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.Optional;

public class BroadcastMessage implements Storable {
    public final PublicKey key;
    public final SHA256Hash hash;

    public BroadcastMessage(PublicKey key, SHA256Hash hash) {
        this.key = key;
        this.hash = hash;
    }

    public static Optional<BroadcastMessage> load(InputStream is) {
        return PublicKey.loadFrom(is).flatMap(kk ->
                        SHA256Hash.loadFrom(is).map(hh -> new BroadcastMessage(kk, hh))
        );
    }

    @Override
    public boolean store(OutputStream os) {
        return key.store(os) && hash.store(os);
    }

    public static int byteLength() {
        return PublicKey.byteLength() + SHA256Hash.byteLength();
    }
}
